File: backend/app/schemas/issue.py
"""Pydantic v2 schemas for Issues."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class IssueDomain(str, Enum):
    AI_ML = "AI/ML"
    DATA = "Data"
    WEB = "Web"
    CLOUD_DEVOPS = "Cloud/DevOps"
    SECURITY = "Security"
    SYSTEMS = "Systems"


class IssueDifficulty(str, Enum):
    EASY = "Easy"
    MEDIUM = "Medium"
    HARD = "Hard"


class LabelSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    color: str = "6b7280"
    description: Optional[str] = None


class IssueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="Unique composite identifier: owner/repo#number")
    repo_owner: str
    repo_name: str
    issue_number: int
    title: str
    body: Optional[str] = ""
    # Present only when the original body exceeded the LLM synthesis cap and a
    # precomputed summary was successfully stored. The raw `body` remains authoritative.
    body_summary: Optional[str] = None
    html_url: str
    author: str = "unknown"
    domain: IssueDomain
    tech_stack: List[str] = Field(default_factory=list)
    difficulty: IssueDifficulty
    estimated_hours: float
    has_bounty: bool = False
    bounty_amount_usd: Optional[float] = None
    bounty_source: Optional[str] = None
    bounty_url: Optional[str] = None
    hourly_roi: Optional[float] = None
    state: str = "open"
    comments_count: int = 0
    labels: List[Dict[str, Any]] = Field(default_factory=list)
    github_created_at: datetime
    github_updated_at: datetime
    indexed_at: datetime


class PaginatedIssuesResponse(BaseModel):
    items: List[IssueResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/issue.ts
export type Domain = 'ai_ml' | 'data' | 'web' | 'cloud_devops' | 'security' | 'systems';

export type Difficulty = 'easy' | 'medium' | 'hard' | 'good_first_issue' | 'intermediate' | 'advanced';

export type BountySource = 'polar' | 'algora' | 'github_sponsors' | 'issuehunt' | 'custom';

export type ViewMode = 'grid' | 'table' | 'compact';

export interface Repository {
  id?: string;
  name: string;
  owner: string;
  stars: number;
  forks: number;
  language: string;
  avatarUrl?: string;
  repoUrl: string;
  description?: string;
}

export interface Bounty {
  id?: string;
  amountUsd: number;
  currency: string;
  source: BountySource;
  sourceUrl: string;
  isFunded: boolean;
  claimed?: boolean;
}

export interface Issue {
  id: string; // e.g. "vllm-project/vllm#4928"
  githubIssueNumber: number;
  title: string;
  body: string;
  /**
   * LLM-condensed body, present only when the original description exceeded the
   * backend cap (LLM_BODY_MAX_CHARS). When set, it is a faithful <8000-char
   * summary of `body`; when absent, `body` is already the full description.
   */
  bodySummary?: string;
  issueUrl: string;
  repository: Repository;
  domain: Domain;
  difficulty: Difficulty;
  estimatedMinutesToSolve: number;
  effortHours?: number;
  techStack: string[];
  bounty?: Bounty;
  hourlyRoiUsd?: number;
  confidenceScore: number;
  createdAt: string;
  updatedAt: string;
  hasTriageReport: boolean;
  labels?: string[];
  author?: string;
}

export interface PaginatedIssuesResponse {
  items: Issue[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
  domainCounts?: Record<string, number>;
  totalBountyPoolUsd?: number;
  /** True when this payload is offline demo/sample data, not the live backend stream. */
  isDemo?: boolean;
}

export interface FilterState {
  domain: Domain | 'all';
  difficulty: Difficulty | 'all';
  techStack: string[];
  hasBountyOnly: boolean;
  minBounty: number;
  timeToSolve: 'all' | 'lt_30m' | '30m_2h' | '2h_6h' | 'gt_6h';
  search: string;
  sortBy: 'created_desc' | 'bounty_desc' | 'roi_desc' | 'time_asc' | 'confidence_desc';
  page: number;
  pageSize: number;
  viewMode: ViewMode;
}
